Update text_similarity_rank_retriever to support inference ID as an argument when reranking on chunks

mridula-s109 opened the pull request:
Adding support to automatically use the best chunking size based on the input 'inference_id' provided.

Tested these below scenarios
```
# Full chunking control
POST my-index/_search
{
  "retriever": {
    "text_similarity_reranker": {
      "retriever": { "standard": { "query": { "match": { "text": "query" } } } },
      "field": "text",
      "inference_id": ".rerank-v1-elasticsearch",
      "inference_text": "search query",
      "chunk_rescorer": {
        "size": 3,
        "chunking_settings": {
          "strategy": "sentence",
          "max_chunk_size": 50,
          "sentence_overlap": 0
        }
      },
      "rank_window_size": 10
    }
  }
}

# Partial chunking (only max_chunk_size)
POST my-index/_search
{
  "retriever": {
    "text_similarity_reranker": {
      "retriever": { "standard": { "query": { "match": { "text": "query" } } } },
      "field": "text",
      "inference_id": ".rerank-v1-elasticsearch",
      "inference_text": "search query",
      "chunk_rescorer": {
        "size": 3,
        "chunking_settings": {
          "max_chunk_size": 100
        }
      },
      "rank_window_size": 10
    }
  }
}

# Auto-resolve to defaults
POST my-index/_search
{
  "retriever": {
    "text_similarity_reranker": {
      "retriever": { "standard": { "query": { "match": { "text": "query" } } } },
      "field": "text",
      "inference_id": ".rerank-v1-elasticsearch",
      "inference_text": "search query",
      "chunk_rescorer": {
        "size": 3
      },
      "rank_window_size": 10
    }
  }
}

# No chunking
POST my-index/_search
{
  "retriever": {
    "text_similarity_reranker": {
      "retriever": { "standard": { "query": { "match": { "text": "query" } } } },
      "field": "text",
      "inference_id": ".rerank-v1-elasticsearch",
      "inference_text": "search query",
      "rank_window_size": 10
    }
  }
}
```

mridula-s109 committed:
Accept inference id in parser

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Wrote the functionality

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

kderusso commented on server/src/main/java/org/elasticsearch/TransportVersions.java:
```diff
@@ -98,6 +98,7 @@ static TransportVersion def(int id) {
     public static final TransportVersion INITIAL_ELASTICSEARCH_8_17_4 = def(8_797_0_04);
     public static final TransportVersion INITIAL_ELASTICSEARCH_8_17_5 = def(8_797_0_05);
     public static final TransportVersion V_8_18_0 = def(8_840_0_00);
+    public static final TransportVersion V_8_19_0 = def(8_900_0_00);
```
Adding a new transport version is not necessary for this PR, but for future notice this would not have been the right thing to do. First we're adding a new version for 8.19 while developing for 9.3+. Going forward please reference the updated guidance for versioning here: https://github.com/elastic/elasticsearch/blob/main/docs/internal/Versioning.md

For this PR however, please remove.

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/ChunkScorerConfig.java:
```diff
@@ -52,16 +54,26 @@ public static ChunkingSettings chunkingSettingsFromMap(Map<String, Object> map)
     public ChunkScorerConfig(StreamInput in) throws IOException {
         this.size = in.readOptionalVInt();
         this.inferenceText = in.readString();
+        if (in.getTransportVersion().onOrAfter(TransportVersions.V_8_19_0)) {
```
Remember our discussion about usability here? We should re-use the `inferenceId` specified in the retriever already. Nothing should be added to the chunk scoring config.

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankRetrieverBuilder.java:
```diff
@@ -55,6 +55,7 @@ public class TextSimilarityRankRetrieverBuilder extends CompoundRetrieverBuilder
     public static final ParseField CHUNK_RESCORER_FIELD = new ParseField("chunk_rescorer");
     public static final ParseField CHUNK_SIZE_FIELD = new ParseField("size");
     public static final ParseField CHUNKING_SETTINGS_FIELD = new ParseField("chunking_settings");
+    public static final ParseField CHUNK_SCORER_INFERENCE_ID_FIELD = new ParseField("inference_id");
```
Please remove as we want to use `INFERENCE_ID_FIELD`

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -61,12 +62,15 @@ public TextSimilarityRankFeaturePhaseRankCoordinatorContext(
+    /**
+     * Creates a scoring listener that uses the resolved chunking config for proper chunking behavior.
+     */
+    private ActionListener<InferenceAction.Response> createScoringListener(
```
This refactoring seems like it makes sense at a high level, but I think breaking out the action listener creation into a private method makes this more opaque. 

Furthermore, we need to be sensitive to the fact that chunking is not the default action, it does not happen with every request. We should only pull chunking settings from the inference endpoint if it does exist otherwise it's a wasted call.

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -133,6 +141,22 @@ protected void computeScores(RankFeatureDoc[] featureDocs, ActionListener<float[
+            // Resolve chunking settings if missing
+            ChunkScorerConfig resolvedChunkScorerConfig = resolveChunkingSettings(r);
+            if (resolvedChunkScorerConfig == null) {
+                l.onFailure(new IllegalArgumentException("Failed to resolve chunking settings for chunk_rescorer"));
```
This means that everyone using semantic reranking as-is without chunking rescoring would break!

kderusso reviewed (changes requested):
I performed a very high level review of the approach. Please update this PR with the suggested changes.

mridula-s109 committed:
Removed unnecessary transport versioning

mridula-s109 committed:
Modified chunk scorer to allow null

mridula-s109 committed:
Fixed compile error

mridula-s109 committed:
Merge conflict resolved

mridula-s109 committed:
Merge conflict

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Unnecesary change reverted

mridula-s109 committed:
Nit - removed white space

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Fixed comments

mridula-s109 committed:
Fixed null pointer

mridula-s109 committed:
Fixed a NPE possibility

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Unit tests added

mridula-s109 committed:
comitting to trigger a run

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

Commit:
[CI] Auto commit changes from spotless

mridula-s109 commented:
I am still unable to run the yaml tests locally, but opening up for early feedback!

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Update docs/changelog/137397.yaml

mridula-s109 committed:
All yaml pass

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
fixed the redundant code

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRankFeaturePhaseRankCoordinatorContext.java:
```diff
@@ -235,4 +239,36 @@ private static float normalizeScore(float score) {
+        );
+    }
+
+    private ChunkingSettings extractChunkingSettingsFromResponse(GetInferenceModelAction.Response response) {
```
This method will not work. If you step through the code, you'll see that the Elastic Reranker endpoint does not have chunking settings attached to it. So at the end of the day, all you're doing is returning the current hard coded chunking settings, and we're not actually doing anything with this change. 

If you look at the PR I linked to from the issue: https://github.com/elastic/elasticsearch/pull/132169 this is how to get the actual token size we want to use for Reranker endpoints. You want to use the `GetRerankerWindowSizeAction` here, _not_ `GetInferenceModelAction`.

kderusso reviewed (changes requested):
First pass review, this needs significant work, and you need to step through the code and understand what is happening here when you're requesting reranker size data. 

Furthermore, none of the tests that you added are actually testing the functionality you're trying to implement here. 

Please step through this with a debugger so you can understand exactly what is happening here and what needs to be changed.

mridula-s109 committed:
Reworking

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

Commit:
[CI] Update transport version definitions

kderusso commented on x-pack/plugin/inference/src/yamlRestTest/resources/rest-api-spec/test/inference/70_text_similarity_rank_retriever.yml:
```diff
@@ -984,3 +984,205 @@ setup:
+  - contains: { profile.shards.0.searches.0.query.0.description: "\"strategy\":\"sentence\"" }
+
+---
+"Test chunk rescorer empty object defaults to size=1 and inference endpoint window":
```
Where is the inference window check here?

kderusso reviewed (commented):
Nice progress here @mridula-s109 , there's still some work especially on tests to go. I'd like to see some unit tests surrounding your logic around chunking settings resolution as well.

mridula-s109 committed:
worked on pr comments

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/TextSimilarityRerankingRankFeaturePhaseRankShardContext.java:
```diff
@@ -34,8 +34,16 @@ public class TextSimilarityRerankingRankFeaturePhaseRankShardContext extends Rer
+        if (chunkScorerConfig != null) {
+            this.chunkingSettings = (chunkScorerConfig.chunkingSettings() != null)
+                ? chunkScorerConfig.chunkingSettings()
+                : ChunkScorerConfig.defaultChunkingSettings(ChunkScorerConfig.DEFAULT_CHUNK_SIZE);
```
You raise a good point. When the model window is larger (e.g Cohere at 2800)
the shard still chunks at 300 because it runs before the coordinator can fetch 
the window size. 

This means:
Chunks are smaller than optimal for large-window models but requests always succeed (300 ≤ all supported windows) and no risk of truncation.

To achieve truly optimal chunking (using model's actual window), we'd need to fetch the window size BEFORE shard dispatch - a pipeline refactor that's out of scope here. Happy to open a follow-up issue if you'd like.

mridula-s109 replied:
have resolved the issue with chunking never happening and have also performed the test for jina ai and validated that it works as intended now.

mridula-s109 committed:
Add unit tests for ChunkScorerConfig chunking settings resolution

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
CI failure tackled

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
failing test

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Delete .es

mridula-s109 committed:
Fixed the yaml failure

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

kderusso commented on x-pack/plugin/inference/src/main/java/org/elasticsearch/xpack/inference/rank/textsimilarity/ChunkScorerConfig.java:
```diff
@@ -29,7 +29,7 @@ public class ChunkScorerConfig implements Writeable, ToXContentObject {
     public static final int DEFAULT_CHUNK_SIZE = 300;
```
We should remove `DEFAULT_CHUNK_SIZE` as it is hard coded and not applicable to all models.

mridula-s109 replied:
have done the same!

kderusso commented on x-pack/plugin/inference/src/test/java/org/elasticsearch/xpack/inference/rank/textsimilarity/ChunkScorerConfigTests.java:
```diff
@@ -0,0 +1,108 @@
+        assertEquals("sentence", settings.asMap().get("strategy"));
+    }
+
+    public void testDefaultChunkingSettingsWithNull() {
```
This test should not pass, because the chunk size should be pulled from the inference endpoint!

kderusso reviewed (changes requested):
High level functional review, still needs to be addressed.

mridula-s109 committed:
Modified files to make sure the feature is working as intended with the right chunking size

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

Commit:
[CI] Auto commit changes from spotless

mridula-s109 committed:
Refactored/cleanedup code a bit

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 commented:
@kderusso Thanks for your patience! I have addressed the comments and also made sure the functionality is working as intended for different reranking models. Please let me know if there are any concerns or optimisation needed.

kderusso reviewed (approved):
Looks good, thanks for all your hard work iterating! 

Can you please also update https://github.com/elastic/elasticsearch/blob/main/docs/reference/elasticsearch/rest-apis/retrievers/text-similarity-reranker-retriever.md to say that we default to chunking settings that will fit into the model associated with `inference_id`'s token window? (This can be done as a followup if you want).

mridula-s109 committed:
Updated docs

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Updated docs

mridula-s109 committed:
Added applies to

mridula-s109 committed:
Applies to tag

kderusso commented on docs/reference/elasticsearch/rest-apis/retrievers/text-similarity-reranker-retriever.md:
```diff
@@ -101,9 +101,9 @@ score = ln(score), if score < 0
-    Settings for chunking text into smaller passages for scoring and reranking. Defaults to the optimal chunking settings for [Elastic Rerank](docs-content:///explore-analyze/machine-learning/nlp/ml-nlp-rerank.md). Refer to the [Inference API documentation](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put#operation-inference-put-body-application-json-chunking_se ...
-    :::{warning} 
-    If you configure chunks larger than the reranker's token limit, the results may be truncated. This can degrade relevance significantly.
+    Settings for chunking text into smaller passages for scoring and reranking. {applies_to}`stack: ga 9.4` By default, chunking settings are configured to fit within the token window of the model associated with the `inference_id`. Refer to the [Inference API documentation](https://www.elastic.co/docs/api/doc/elasticsearch/operation/operation-inference-put#operation-inference-put-body-applicatio ...
```
Can we add beta 9.2 here? Otherwise LGTM thanks - no need to re-review on my end other than that.

mridula-s109 replied:
Done, thankyou :)

mridula-s109 committed:
Updated doc to 9.2

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity

mridula-s109 committed:
Merge branch 'main' into add-inferenceid-support-textsimilarity
